cmd/download.go

```diff
@@ -0,0 +1,250 @@
+
+		procStart := time.Now()
+		responseBody, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return nil, fmt.Errorf("could not read FHIR server response: %v\n", err)
+		}
+		response.Body.Close()
+		stats.totalBytesIn += int64(len(responseBody))
+
+		var resourceBundle map[string]interface{}
+		err = json.Unmarshal(responseBody, &resourceBundle)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse FHIR server response: %v\n", err)
+		}
+
+		numResources, err := writeResources(resourceBundle, sink)
+		if err != nil {
+			return nil, fmt.Errorf("could not write resource to output file: %v\n", err)
+		}
+		stats.resourcesPerPage = append(stats.resourcesPerPage, numResources)
+
+		nextPageURL, err = getNextPageURL(resourceBundle)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse the next page link within the FHIR server response: %v\n", err)
+		}
+		stats.processingDurations = append(stats.processingDurations, time.Since(procStart).Seconds())
+	}
+
+	stats.endTime = time.Now()
+	return &stats, nil
+}
+
+// createOutputFileOrDie creates the output file at the given filepath if it does not already exist
+// and returns the file handle.
+// This is a non-destructive operation. Hence, if a file already exists at the given filepath then
+// the command exits with a non-success error code. If any other error case the command exits with
+// a non-success error code as well.
+//
+// Note: The callee has to make sure that the file handle is closed properly.
+func createOutputFileOrDie(filepath string) *os.File {
+	outputFile, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+	if err != nil {
+		if os.IsExist(err) {
+			fmt.Printf("The output file at `%v` does already exist.\n", filepath)
+			os.Exit(3)
+		} else {
+			fmt.Printf("could not open/create the output file at `%s`: %v", filepath, err)
+			os.Exit(4)
+		}
+	}
+	return outputFile
+}
+
+// writeResources writes all resources of the given resourceBundle as-is to the given sink.
+// Returns the number of written resources or an error if the resources can not be written to the sink.
+// An error is only possible if there is an actual issue writing to the file or the given resource
+// bundle is invalid with regards to the FHIR specification.
+func writeResources(resourceBundle map[string]interface{}, sink io.Writer) (resources int, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("encountered a malformed resource bundle")
+		}
+	}()
+
+	entries := resourceBundle["entry"].([]interface{})
+	for e := range entries {
+		entry := entries[e].(map[string]interface{})
+
+		if entry["resource"] != nil {
+			resources++
+			jsonEncoded, err := json.Marshal(entry["resource"])
+			if err != nil {
+				return -1, fmt.Errorf("could not encode resource as JSON: %v\n", err)
+			}
+
+			if _, err = sink.Write(append(jsonEncoded, '\n')); err != nil {
```

Comment:
I would write the newline seperately. because otherwise the slice of chars has to be resized - right?

Reply:
The resize could potentially happen if the slice's content already occupies as much space as the slice itself has capacity.

However, I would not trade this for 2 writes. The drawback is that we would have to deal with error checking twice and also access the disk twice. A potential resize should be faster than those accesses.

Reply:
Isn't the writer buffered?

Reply:
Have been digging a bit on this one. The write to the file itself is not buffered on an application level since a `write` syscall is triggered directly on the file descriptor. However, the underlying file system COULD have a buffer system (deferred writes) in place.

So I would argue that this is pretty much use case dependent. If a single FHIR resource can grow fairly large in size then two separate writes may be better in terms of memory usage. Since I don't have any information on this I will go for the 2 writes just to be safe.

A call to `Sync` will also get added right before closing the file to make sure data is definitely written to the device.

Reply:
ok